app/models/principal.rb

```diff
@@ -7,7 +7,8 @@ class Principal < ActiveRecord::Base
   has_one :firm,
     -> { where(parent_id: nil) },
     primary_key: :fca_number,
-    foreign_key: :fca_number
+    foreign_key: :fca_number,
+    dependent: :destroy
```

Comment:
The parent firm is already configured to cascade delete subsidiary (trading name) firms, so all we have to do is delete the primary firm associated with the principal to get the effect we want.